examples/user_input.rs

```diff
@@ -47,10 +51,67 @@ impl Default for App {
             input: String::new(),
             input_mode: InputMode::Normal,
             messages: Vec::new(),
+            cursor_in_input: 0,
         }
     }
 }
 
+impl App {
+    fn move_cursor_left(&mut self) {
+        if !self.is_cursor_leftmost() {
+            let cursor_moved_left = self.cursor_in_input - 1;
+            self.cursor_in_input = cursor_moved_left;
+        };
+    }
+
+    fn move_cursor_right(&mut self) {
+        if !self.is_cursor_rightmost() {
+            let cursor_moved_right = self.cursor_in_input + 1;
+            self.cursor_in_input = cursor_moved_right;
+        };
+    }
+
+    fn enter_char_at_cursor(&mut self, new_char: char) {
+        if self.is_cursor_leftmost() {
+            self.input.push(new_char);
+        } else {
+            self.input.insert(self.cursor_in_input, new_char);
+        }
+
+        self.move_cursor_right();
+    }
+
+    fn delete_char_at_cursor(&mut self) {
+        if !self.is_cursor_leftmost() {
+            let current_index = self.cursor_in_input;
+            let from_left_to_current_index = current_index - 1;
+
+            let before_char_to_delete = self.input.chars().take(from_left_to_current_index);
+            let after_char_to_delete = self.input.chars().skip(current_index);
+
+            self.input = before_char_to_delete.chain(after_char_to_delete).collect();
+
+            self.move_cursor_left();
+        }
+    }
+
+    fn is_cursor_leftmost(&self) -> bool {
+        self.cursor_in_input == 0
+    }
+    fn is_cursor_rightmost(&self) -> bool {
+        self.cursor_in_input == self.input.len()
+    }
+
+    fn reset_cursor(&mut self) {
+        self.cursor_in_input = 0;
+    }
+
+    fn submit_message(&mut self) {
+        self.messages.push(self.input.drain(..).collect());
```

Comment:
I don't understand why you're pushing the message like this, why not simply push the string ?

Reply:
I copied it from the earlier version. I agree that just pushing the string is simpler and better.

Reply:
As this is an example I'd suggest just cloning it. Wouldn't the `.collect()` allocate in this case anyways ?

Reply:
Since clearning is needed, I would offer this version 

```rs
fn submit_message(&mut self) {
    self.messages.push(self.input.clone());
    self.input.clear();
    self.reset_cursor();
}
```

If you like, I apply this version.

Reply:
Applied the offered change.